runtime/compiler/control/JITServerHelpers.cpp

```diff
@@ -44,84 +47,309 @@ TR::Monitor *JITServerHelpers::_clientStreamMonitor = NULL;
-   for (size_t i = 0; i < origRomClass->romMethodCount; ++i)
+   auto name = J9ROMCLASS_CLASSNAME(romClass);
+   // Primitive ROMClasses have different layout (see runtime/vm/romclasses.c): the last
+   // ROMClass includes all the others' UTF8 name strings in its romSize, which breaks the
+   // generic packing implementation. Pretend that its romSize only includes the header.
+   size_t origRomSize = J9ROMCLASS_IS_PRIMITIVE_TYPE(romClass) ? sizeof(*romClass) : romClass->romSize;
+   size_t totalSize = origRomSize;
+
+   // Remove intermediate class data (not used by JIT)
+   uint8_t *icData = J9ROMCLASS_INTERMEDIATECLASSDATA(romClass);
+   if (JITServerHelpers::isAddressInROMClass(icData, romClass) && (icData != (uint8_t *)romClass))
       {
-      totalSize += getPackedUTF8Size(J9ROMMETHOD_NAME(romMethod), origRomClass);
-      totalSize += getPackedUTF8Size(J9ROMMETHOD_SIGNATURE(romMethod), origRomClass);
-      romMethod = nextROMMethod(romMethod);
+      TR_ASSERT_FATAL(icData + romClass->intermediateClassDataLength == (uint8_t *)romClass + romClass->romSize,
+                      "Intermediate class data not stored at the end of ROMClass %.*s", name->length, name->data);
+      totalSize -= romClass->intermediateClassDataLength;
       }
 
-   // Check if the contiguous part of the ROMClass already contains all the strings visited above
-   if (totalSize == origRomClass->romSize)
-      return std::string((char *)origRomClass, origRomClass->romSize);
+   // All allocated memory is only used in this function
+   TR::StackMemoryRegion stackMemoryRegion(*trMemory);
+   ROMClassPackContext ctx(trMemory, origRomSize);
 
-   J9ROMClass *romClass = (J9ROMClass *)trMemory->allocateHeapMemory(totalSize);
-   if (!romClass)
+   size_t copySize = 0;
+   if (isArrayROMClass(romClass))
+      {
+      copySize = sizeof(*romClass);
+      totalSize = getArrayROMClassPackedSize(romClass);
+      }
+   else
+      {
+      // 1st pass: iterate all strings in the ROMClass to compute its total size (including
+      // interned strings) and map the strings to their locations in the packed ROMClass
+      ctx.callback = sizeInfoCallback;
+      allSlotsInROMClassDo(romClass, slotCallback, NULL, NULL, &ctx);
+      // Handle the case when all strings are interned
+      auto classEnd = (const uint8_t *)romClass + totalSize;
+      ctx.utf8SectionStart = std::min(ctx.utf8SectionStart, classEnd);
+
+      auto end = ctx.utf8SectionEnd ? ctx.utf8SectionEnd : classEnd;
+      TR_ASSERT_FATAL(ctx.utf8SectionSize == end - ctx.utf8SectionStart,
+                      "Missed strings in ROMClass %.*s UTF8 section: %zu != %zu",
+                      name->length, name->data, ctx.utf8SectionSize, end - ctx.utf8SectionStart);
+      end = (const uint8_t *)OMR::alignNoCheck((uintptr_t)end, sizeof(uint64_t));
+      TR_ASSERT_FATAL(end == classEnd, "UTF8 section not stored at the end of ROMClass %.*s: %p != %p",
```

Comment:
Why does `end` need to be the same as `classEnd`?
`end` is what we computed with interned and non-interned strings.
`classEnd` is `romClass+romClass->romSize` which I thought it did not include the interned strings

Reply:
`end` is computed with non-interned strings only.

Reply:
Right